NativeAuthSampleApp/MultiFactorAuthenticationViewController.swift

```diff
@@ -0,0 +1,333 @@
+            showResultText("Invalid username or password")
+        } else {
+            showResultText("Error while signing in: \(error.errorDescription ?? "No error description")")
+        }
+    }
+
+    func onSignInAwaitingMFA(newState: AwaitingMFAState) {
+        print("SignInStartDelegate: onSignInAwaitingMFA: \(newState)")
+
+        let alert = UIAlertController(title: "MFA required", message: "Do you want to proceed with MFA?", preferredStyle: .alert)
+
+        alert.addAction(UIAlertAction(title: "OK", style: .default, handler: { _ in
+            newState.requestChallenge(delegate: self)
+        }))
+
+        alert.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: { _ in
+            self.resultTextView.text = "Second factor authentication required"
+        }))
+
+        present(alert, animated: true)
+    }
+}
+
+// MARK: MFARequestChallengeDelegate
+
+extension MultiFactorAuthenticationViewController: MFARequestChallengeDelegate {
+    
+    func onMFARequestChallengeError(error: MFAError, newState: MFARequiredState?) {
+        print("MFARequestChallengeDelegate: onMFARequestChallengeError: \(error)")
+        showResultText("Unexpected error while requesting challenge")
+        dismissVerifyCodeModal()
+    }
+
+    func onMFARequestChallengeVerificationRequired(
+        newState: MFARequiredState,
+        sentTo: String,
+        channelTargetType: MSALNativeAuthChannelType,
+        codeLength: Int
+    ) {
+        print("MFARequestChallengeDelegate: onMFARequestChallengeVerificationRequired: \(newState)")
+
+        guard verifyCodeViewController == nil else {
+            return
+        }
+
+        showVerifyCodeModal(submitCallback: { [weak self] code in
+                                guard let self = self else { return }
+
+                                newState.submitChallenge(challenge: code, delegate: self)
+                            },
+                            resendCallback: { [weak self] in
+                                guard let self = self else { return }
+
+                                newState.requestChallenge(delegate: self)
+                            }, cancelCallback: { [weak self] in
+                                guard let self = self else { return }
+
+                                self.dismissVerifyCodeModal()
+                                showResultText("Action cancelled")
+                            })
+    }
+}
+
+// MARK: MFASubmitChallengeDelegate
+
+extension MultiFactorAuthenticationViewController: MFASubmitChallengeDelegate {
+
+    func onMFASubmitChallengeError(error: MFASubmitChallengeError, newState: MFARequiredState?) {
+        print("MFASubmitChallengeDelegate: onMFASubmitChallengeError: \(error)")
+
+        if error.isInvalidChallenge {
+            guard let newState = newState else {
+                print("Unexpected state. Received invalidCode but newState is nil")
```

Comment:
Can we just say: "expected new state not nil"? With the current message it seems that the user needs to expect newState not nil when error is invalidChallenge

Reply:
I think the message is clear.

> With the current message it seems that the user needs to expect newState not nil when error is invalidChallenge

In fact this is what they should expect.